fem/pbilinearform.hpp

```diff
@@ -150,6 +150,16 @@ class ParBilinearForm : public BilinearForm
        vectors on the true dofs. */
    void TrueAddMult(const Vector &x, Vector &y, const double a = 1.0) const;
 
+   using BilinearForm::InnerProduct;
+   /// Compute inner product x^t A y (grid function version)
+   /** @note It is assumed that the parallel system matrix is assembled,
+       see FormSystemMatrix(). */
+   double InnerProduct(const ParGridFunction &x, const ParGridFunction &y);
+   /// Compute inner product x^t A y (vector version)
+   /** @note It is assumed that the parallel system matrix is assembled,
+       see FormSystemMatrix(). */
+   double InnerProduct(HypreParVector &x, HypreParVector &y);
```

Comment:
I have a few comments/objections with regards to these two `InnerProduct` functions. 
1. A `ParGridFunction` operates on `ldofs` therefore it's missleading to actually perform the dot product on true dofs. Maybe a better name would be `TrueInnerProduct`. This however can be done by the existing method `void ParBilinearForm::TrueAddMult`.
2. More importantly, the naming "InnerProduct" implies a positive definite and symmetric form which is not always true.

Reply:
1. I would not say that `ParGridFunction` solely operates on ldofs. For that, you could use just plain `GridFunction`, but `ParGridFunction` represents a global thing. However, it is assumed most of the time that the grid function is "distributed" through `ParGridFunction::Distribute()`, i.e., it is a prolonged T-vector, so you may easily use the L-vector. Anyway, I think it is probably the first thing one would expect there (but somewhat dangerous), so I would agree with your suggestion. The version based on the local inner product would here instead.
2. Well, you are mathematically right, but it is just a name used throughout MFEM. It is  on user to apply it on appropriate matrices. If you really want, I may add a note there...

Reply:
1. Ok, I just implemented your suggestion, so there is `InnerProduct` and `TrueInnerProduct`, where the user is warned about the necessary consistency of the L-vectors in the former.
2. Regarding the second comment, that is bit tricky. A note in the code about the mathematical requirements on the forms would help beginners definitely, but it would confuse experts, because these properties are not guaranteed on the numerical level. Therefore, I went a way around and followed the description from `BilinearForm`, where it is simply not stated in the description that it is an inner product and just the applied formula appears there. (For consistency, I swapped the arguments.)